src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleEnumerable.cs

```diff
@@ -48,16 +48,36 @@ public static partial class JsonSerializer
 
             if (state.Current.Enumerator != null && state.Current.Enumerator.MoveNext())
             {
+                // If the enumerator contains typeof(object), get the run-time type
+                if (elementClassInfo.ClassType == ClassType.Object && jsonPropertyInfo.ElementClassInfo.Type == typeof(object))
+                {
+                    object currentValue = state.Current.Enumerator.Current;
+                    if (currentValue != null)
+                    {
+                        Type runtimeType = currentValue.GetType();
+                        
+                        // Ignore object() instances since they are handled as an empty object.
+                        if (runtimeType != typeof(object))
+                        {
+                            elementClassInfo = options.GetOrAddClass(runtimeType);
+                        }
+                    }
+                }
+
                 if (elementClassInfo.ClassType == ClassType.Value)
                 {
                     elementClassInfo.GetPolicyProperty().WriteEnumerable(options, ref state.Current, ref writer);
                 }
+                else if (state.Current.Enumerator.Current == null)
```

Comment:
> Fix issue with null object in array

What does the test coverage look like now? Any other missing branches worth testing?